### PR TITLE
[TextFields] Converting NSCoding to NSSecureCoding

### LINKED ISSUE #2837: MDCTextInputCommonFundament does not conform to NSSecureCoding
Source:
https://github.com/material-components/material-components-ios/blob/develop/components/TextFields/src/private/MDCTextInputCommonFundament.m

NSSecureCoding documentation:
https://developer.apple.com/documentation/foundation/nssecurecoding?language=objc
http://nshipster.com/nssecurecoding/

### LINKED ISSUE #2838: MDCTextInputUnderlineView does not conform to NSSecureCoding
Source:
https://github.com/material-components/material-components-ios/blob/develop/components/TextFields/src/MDCTextInputUnderlineView.m

NSSecureCoding documentation:
https://developer.apple.com/documentation/foundation/nssecurecoding?language=objc
http://nshipster.com/nssecurecoding/

### PR DESCRIPTION
**Pull request by willlarche**
Closes #2837  
Closes #2838

### CODE REVIEW

File: components/TextFields/src/MDCTextInputControllerFullWidth.m [hunk 1]
```diff
@@ -147,16 +148,23 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
...
+        [aDecoder decodeObjectOfClass:[UIColor class]
+                               forKey:MDCTextInputControllerFullWidthInlinePlaceholderColorKey];
     _inlinePlaceholderFont =
-        [aDecoder decodeObjectForKey:MDCTextInputControllerFullWidthInlinePlaceholderFontKey];
-    _textInput = [aDecoder decodeObjectForKey:MDCTextInputControllerFullWidthTextInputKey];
+        [aDecoder decodeObjectOfClass:[UIFont class]
+                               forKey:MDCTextInputControllerFullWidthInlinePlaceholderFontKey];
+    _textInput =
+        [aDecoder decodeObjectOfClass:[UIView<MDCTextInput> class]
+                               forKey:MDCTextInputControllerFullWidthInlinePlaceholderFontKey];
```
Comment (yarneo) [inline components/TextFields/src/MDCTextInputControllerFullWidth.m]: This needs to be `MDCTextInputControllerFullWidthTextInputKey`

File: components/TextFields/src/MDCTextInputControllerLegacyDefault.m [hunk 2]
```diff
@@ -218,22 +221,31 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
...
     _inlinePlaceholderColor =
-        [aDecoder decodeObjectForKey:MDCTextInputControllerLegacyDefaultInlinePlaceholderColorKey];
+        [aDecoder decodeObjectOfClass:[UIColor class]
+                               forKey:MDCTextInputControllerLegacyDefaultInlinePlaceholderColorKey];
     _leadingUnderlineLabelTextColor = [aDecoder
-        decodeObjectForKey:MDCTextInputControllerLegacyDefaultLeadingUnderlineLabelTextColor];
-    _normalColor = [aDecoder decodeObjectForKey:MDCTextInputControllerLegacyDefaultNormalColorKey];
-    _textInput = [aDecoder decodeObjectForKey:MDCTextInputControllerLegacyDefaultTextInputKey];
+        decodeObjectOfClass:[UIColor class]
+                     forKey:MDCTextInputControllerLegacyDefaultFloatingPlaceholderNormalColorKey];
```
Comment (yarneo) [inline components/TextFields/src/MDCTextInputControllerLegacyDefault.m]: This needs to be `MDCTextInputControllerLegacyDefaultLeadingUnderlineLabelTextColor`

**Review by yarneo — COMMENTED** [review]
Please see comments and reverify that all the keys are the right keys. Otherwise looks good!

File: components/TextFields/src/MDCTextInputControllerFullWidth.m (hunk 1 above)
Comment (willlarche) [reply components/TextFields/src/MDCTextInputControllerFullWidth.m]: Oops! Fixed these two. My eyes must have started to glaze like a ham at this point. PTAL

File: components/TextFields/src/MDCTextInputControllerBase.m [hunk 3]
```diff
@@ -211,34 +211,50 @@ - (instancetype)init {
...
+    _disabledColor = [aDecoder decodeObjectOfClass:[UIColor class]
+                                            forKey:MDCTextInputControllerBaseDisabledColorKey];
+    _errorColor = [aDecoder decodeObjectOfClass:[UIColor class]
+                                         forKey:MDCTextInputControllerBaseErrorColorKey];
     _expandsOnOverflow = [aDecoder decodeBoolForKey:MDCTextInputControllerBaseExpandsOnOverflowKey];
     _floatingEnabled = [aDecoder decodeBoolForKey:MDCTextInputControllerBaseFloatingEnabledKey];
     _floatingPlaceholderNormalColor =
-        [aDecoder decodeObjectForKey:MDCTextInputControllerBaseFloatingPlaceholderNormalColorKey];
+        [aDecoder decodeObjectOfClass:[UIColor class]
+                               forKey:MDCTextInputControllerBaseNormalColorKey];
```
Comment (yarneo) [inline components/TextFields/src/MDCTextInputControllerBase.m]: needs to be `MDCTextInputControllerBaseFloatingPlaceholderNormalColorKey`

**Review by yarneo — APPROVED** [review]
Please fix comment and also verify that our unit tests cover archiving and unarchiving the objects we are now using `NSSecureEncoding` for. Approving PR.

File: components/TextFields/src/MDCTextInputControllerBase.m (hunk 3 above)
Comment (willlarche) [reply components/TextFields/src/MDCTextInputControllerBase.m]: Thanks for catching these! It's not easy.

**Pull request merged**